pkg/datapath/linux/probes/probes.go

```diff
@@ -204,13 +199,19 @@ func (p *ProbeManager) SystemKernelHz() (int, error) {
 // SystemConfigProbes performs a check of kernel configuration parameters. It
 // returns an error when parameters required by Cilium are not enabled. It logs
 // warnings when optional parameters are not enabled.
+//
+// When kernel config file is not found, bpftool can't probe kernel configuration
+// parameter real setting, so only return error log when kernel config file exists
+// and kernel configuration parameter setting is disabled
 func (p *ProbeManager) SystemConfigProbes() error {
+	var notFound bool
 	if !p.KernelConfigAvailable() {
-		return ErrKernelConfigNotFound
+		notFound = true
+		log.Warn("Kernel Config file not found: If agent fail to start, Please check kernel requirements https://docs.cilium.io/en/stable/operations/system_requirements")
```

Comment:
Can we achieve the same logic as this PR by *only* adding this warning message?

I found a few things to be confusing in this PR:
- `notFound` then gets checked as `!notFound` below, so then it's confusing to think about what "not not found" double negative means. At the very least this can be `found = false`.
- But then, if all of the checks below are being skipped in the `found == false` case , this could just return `nil`.
- If this line returned `nil` though, it's a bit strange because this function *is* actually having an error condition, it's just that the callers may not want to treat that as a fatal error, they can tolerate the error and continue. This is the case already today.

In the end, it seems like the core difference that this PR proposes is just to avoid logging the message as `BPF system config check: NOT OK.`, this instead logs a more friendly warning. If that's the case, then we don't need to make the logic checks more complicated here, the complete PR could be just to add this one log warning message.

Reply:
If I recall correct, in https://github.com/cilium/cilium/issues/20858#issuecomment-1213342951, in my reproduction, without this PR, the original code does not log anything, not even `BPF system config check: NOT OK`.  I could make this PR just as simple as below, I will have to double check and test again. 

```
diff --git a/pkg/datapath/linux/requirements.go b/pkg/datapath/linux/requirements.go
index 199c135716..b12452550c 100644
--- a/pkg/datapath/linux/requirements.go
+++ b/pkg/datapath/linux/requirements.go
@@ -146,7 +146,7 @@ func CheckMinRequirements() {
                if err := probeManager.SystemConfigProbes(); err != nil {
                        errMsg := "BPF system config check: NOT OK."
                        // TODO(brb) warn after GH#14314 has been resolved
-                       if !errors.Is(err, probes.ErrKernelConfigNotFound) {
+                       if errors.Is(err, probes.ErrKernelConfigNotFound) {
                                log.WithError(err).Warn(errMsg)
                        }
                }
```

Reply:
I just tested above simple change, if I do not have required kernel BPF config, but I have kernel config file found, there will be no any log, just like original issue reported in https://github.com/cilium/cilium/issues/20858, agent crashes, but no early log clue on maybe required kernel BPF config might be missing. I admit the logic code in this PR is kind of confusing, but it does give user useful log clue. see the 4 test scenario with this PR https://github.com/cilium/cilium/pull/20889#issuecomment-1214179412

Reply:
to make the code bit clear, I could rename `notFound` to `found`, that seems less confusing

Reply:
I made the change, I think the logic is clear now :)

Reply:
the  following two `if !kernelOption.Enabled && found` check is to only log when required/optional kernel config not enabled and kernel config file is found, because bpftool probe kernel features based on kernel config file either in `/proc/config.gz` or `/boot/`, if none of these kernel config file exists, bpftool can't probe kernel feature, so bpftool don't really know if required/optional kernel config is enabled or not, log the message without kernel config file found is not correct.

Reply:
> I just tested above simple change, if I do not have required kernel BPF config, but I have kernel config file found, there will be no any log ...

The changes in this function only make a difference in the case `if !p.KernelConfigAvailable()`. I don't understand how you could observe a difference in the case where the kernel config is found.

Reply:
```
diff --git a/pkg/datapath/linux/requirements.go b/pkg/datapath/linux/requirements.go
index 199c135716..b12452550c 100644
--- a/pkg/datapath/linux/requirements.go
+++ b/pkg/datapath/linux/requirements.go
@@ -146,7 +146,7 @@ func CheckMinRequirements() {
                if err := probeManager.SystemConfigProbes(); err != nil {
                        errMsg := "BPF system config check: NOT OK."
                        // TODO(brb) warn after GH#14314 has been resolved
-                       if !errors.Is(err, probes.ErrKernelConfigNotFound) {
+                       if errors.Is(err, probes.ErrKernelConfigNotFound) {
                                log.WithError(err).Warn(errMsg)
                        }
                }
```

This change is wrong, because it hides all of the other errors. If you were to do an approach like this, it would need to add new lines something like this:

                if err := probeManager.SystemConfigProbes(); err != nil {
                        errMsg := "BPF system config check: NOT OK."
                        // TODO(brb) warn after GH#14314 has been resolved
                        if !errors.Is(err, probes.ErrKernelConfigNotFound) {

                                log.WithError(err).Warn(errMsg)
+                       } else if errors.Is(err, probes.ErrKernelConfigNotFound) {                        
                                log.WithError(err).Warn("If cilium-agent fails to start, Please check kernel requirements https://docs.cilium.io/en/stable/operations/system_requirements")}
                }

Reply:
Went to original PR route, but keep this in mind in case original PR is flawed (unlikely) :)